src/plugins/swap.c

```diff
@@ -198,56 +201,150 @@ gboolean bd_swap_swapon (const gchar *device, gint priority, GError **error) {
-        g_io_channel_shutdown (dev_file, FALSE, &tmp_error);
-        g_clear_error (&tmp_error);
+                     "No superblock detected on the device '%s'", device);
         bd_utils_report_finished (progress_id, (*error)->message);
+        blkid_free_probe (probe);
+        close (fd);
         return FALSE;
     }
 
-    g_io_channel_shutdown (dev_file, FALSE, &tmp_error);
-    g_clear_error (&tmp_error);
+    status = blkid_probe_lookup_value (probe, "TYPE", &value, NULL);
+    if (status != 0) {
+        g_set_error (error, BD_SWAP_ERROR, BD_SWAP_ERROR_UNKNOWN_STATE,
+                     "Failed to get filesystem type for the device '%s'", device);
+        bd_utils_report_finished (progress_id, (*error)->message);
+        blkid_free_probe (probe);
+        close (fd);
+        return FALSE;
+    }
 
-    if (g_str_has_prefix (dev_status, "SWAP-SPACE")) {
+    if (g_strcmp0 (value, "swap") != 0) {
+        g_set_error (error, BD_SWAP_ERROR, BD_SWAP_ERROR_UNKNOWN_STATE,
+                     "Device '%s' is not a swap", device);
+        bd_utils_report_finished (progress_id, (*error)->message);
+        blkid_free_probe (probe);
+        close (fd);
+        return FALSE;
+    }
+
+    status = blkid_probe_lookup_value (probe, "SBMAGIC", &value, NULL);
+    if (status != 0) {
+        g_set_error (error, BD_SWAP_ERROR, BD_SWAP_ERROR_UNKNOWN_STATE,
+                     "Failed to get swap status on the device '%s'", device);
+        bd_utils_report_finished (progress_id, (*error)->message);
+        blkid_free_probe (probe);
+        close (fd);
+        return FALSE;
+    }
+
+    if (g_strcmp0 (value, "SWAP-SPACE") == 0) {
         g_set_error (error, BD_SWAP_ERROR, BD_SWAP_ERROR_ACTIVATE,
                      "Old swap format, cannot activate.");
         bd_utils_report_finished (progress_id, (*error)->message);
+        blkid_free_probe (probe);
+        close (fd);
         return FALSE;
-    } else if (g_str_has_prefix (dev_status, "S1SUSPEND") || g_str_has_prefix (dev_status, "S2SUSPEND")) {
+    } else if (g_strcmp0 (value, "S1SUSPEND") == 0 || g_strcmp0 (value, "S2SUSPEND") == 0) {
         g_set_error (error, BD_SWAP_ERROR, BD_SWAP_ERROR_ACTIVATE,
                      "Suspended system on the swap device, cannot activate.");
         bd_utils_report_finished (progress_id, (*error)->message);
+        blkid_free_probe (probe);
+        close (fd);
         return FALSE;
-    } else if (!g_str_has_prefix (dev_status, "SWAPSPACE2")) {
+    } else if (g_strcmp0 (value, "SWAPSPACE2") != 0) {
         g_set_error (error, BD_SWAP_ERROR, BD_SWAP_ERROR_ACTIVATE,
                      "Unknown swap space format, cannot activate.");
         bd_utils_report_finished (progress_id, (*error)->message);
+        blkid_free_probe (probe);
+        close (fd);
         return FALSE;
     }
 
+    status_len = (gint64) strlen (value);
+
+    status = blkid_probe_lookup_value (probe, "SBMAGIC_OFFSET", &value, NULL);
```

Comment:
I think this is the last place the `probe` is needed so it should be free'd here.

Reply:
`value` is owned by the probe, so we can't free it here.